web/modules/custom/joinup_core/src/EventSubscriber/GroupInvitationSubscriber.php

```diff
@@ -0,0 +1,142 @@
+    $events[InvitationEvents::REJECT_INVITATION_EVENT] = ['rejectInvitation'];
+
+    return $events;
+  }
+
+  /**
+   * Accepts invitations to join a group.
+   *
+   * @param \Drupal\joinup_invite\Event\InvitationEventInterface $event
+   *   The event that was fired.
+   */
+  public function acceptInvitation(InvitationEventInterface $event): void {
+    $invitation = $event->getInvitation();
+
+    // Ignore invitations to other content entities.
+    if ($invitation->bundle() !== 'group') {
+      return;
+    }
+
+    $user = $invitation->getRecipient();
+    $group = $invitation->getEntity();
+
+    if ($this->getMembership($group, $user)) {
+      drupal_set_message($this->t('There is no action pending for this user.'));
+      // If a user already has a membership and is here, it means that there was
+      // a leftover pending invitation (the subscriber is called after a pending
+      // invitation is accepted). In that case, delete the invitation. The
+      // reason behind it is that if the invitation is not deleted, it might
+      // falsely seem that the membership exists due to an accepted invitation.
+      $invitation->delete();
+    }
+    else {
+      $membership = $this->membershipManager->createMembership($group, $user);
+      if (!empty($invitation->get('invitation_role')->first())) {
+        $role = $invitation->get('invitation_role')->first()->entity;
+        $membership->addRole($role);
+      }
+      // Disable notifications related to memberships.
+      $membership->skip_notification = TRUE;
+      $membership->setState(OgMembershipInterface::STATE_ACTIVE)->save();
+      $facilitator_id = $group->getEntityTypeId() . '-' . $group->bundle() . '-facilitator';
+      $role_argument = $membership->hasRole($facilitator_id) ? $this->t('facilitator') : $this->t('member');
+      drupal_set_message($this->t('You are now a @role of the "@title" @bundle.', [
+        '@role' => $role_argument,
+        '@title' => $group->label(),
+        '@bundle' => $group->get('rid')->entity->getSingularLabel(),
+      ]));
+    }
+  }
+
+  /**
+   * Rejects invitations to join a group.
+   *
+   * @param \Drupal\joinup_invite\Event\InvitationEventInterface $event
+   *   The event that was fired.
+   */
+  public function rejectInvitation(InvitationEventInterface $event): void {
+    $invitation = $event->getInvitation();
+
+    // Ignore invitations to other content entities.
+    if ($invitation->bundle() !== 'group') {
+      return;
+    }
+
+    $user = $invitation->getRecipient();
+    $group = $invitation->getEntity();
+
+    // If there is already a membership, it means that it has been created
+    // somewhere else in the meantime.
+    if (!empty($this->getMembership($group, $user))) {
+      drupal_set_message($this->t('There is no action pending for this user.'));
+    }
+    else {
+      drupal_set_message($this->t('The invitation has been rejected. Thank you for your feedback.'));
+      $invitation->delete();
```

Comment:
We don't delete the invitation. The invitation status is automatically set to 'rejected' in `InvitationController::updateInvitation()`.

Reply:
This makes it much more complicated.. It needs functional decisions on what happens if an invitation is already rejected :/ whether we keep track of the invitations or overwrite them.. This ticket is going too big :(

Reply:
We need to log somewhere the rejection, we should keep track of it. I don't see a use case right now but imagine a new functional request in the future that allows the user to opt also for "Reject the invitation and prevent future invites". As the invitation entity has an explicit `InvitationInterface::STATUS_REJECTED` status we should use it.

Reply:
Just collect rejections is OK for now. In the future there might be a new option "Reject and prevent future invites to this collection". Then we can use the stored rejections.